composer/utils/data.py

```diff
@@ -93,27 +97,86 @@ def pil_image_collate(batch: List[Tuple[Image.Image, Union[Image.Image, Tensor]]
     return image_tensor, target_tensor
 
 
-def add_dataset_transform(dataset, transform):
-    """Flexibly add a transform to the dataset's collection of transforms.
+def add_dataset_transform(dataset: torch.utils.data.Dataset,
+                          transform: Callable,
+                          location: Union[str, type] = "end",
+                          pre_post: str = "pre") -> Dataset:
+    """Flexibly add a transform to the dataset's collection of transforms. Warning: this
+    function will not behave as expected if a dataset's collection of transforms contains
+    nested torchvision.transforms.Compose (e.g dataset.transforms =
+    transforms.Compose([transform0, transform1, transforms.Compose(...)])). This function
+    should be updated to flatten nested compositions.
 
     Args:
         dataset: A torchvision-like dataset
         transform: Function to be added to the dataset's collection of transforms
+        location (str, type), optional: Where to insert the transform in
+            the sequence of transforms. "end" will append to the end or "start" will
+            insert at index 0, a transform (e.g. torchvision.transforms.ToTensor) will
+            insert pre- or post- the first instance of `location`, determined by the
+            `pre_post` argument. Default = "end". If `location` is not in list of
+            transforms, will append to end.
+        pre_post (str), optional: Whether to insert before ("pre") or after ("post") the
+            transform string passed in 'location'. Ignored if transform = "start" or "end". Default = "pre".
 
     Returns:
         The original dataset. The transform is added in-place.
     """
 
     if not hasattr(dataset, "transform"):
-        raise ValueError(f"Dataset of type {type(dataset)} has no attribute 'transform'. Expected TorchVision dataset.")
+        raise AttributeError(textwrap.dedent(f"""Dataset must have attribute `transform`."""))
```

Comment:
do we need the `textwrap.dedent` here and below when the errors are relatively short?

Reply:
Nope, textwrap.dedent is only for multiline errors.

Reply:
Can we use check to ensure the dataset is a subclass of `torchvision.datasets.VisionDataset` (https://pytorch.org/vision/main/generated/torchvision.datasets.VisionDataset.html)?

e.g.

```python3
if not isinstance(dataset, torchvision.datasets.VisionDataset):
    raise TyepError(...)
```

Reply:
Since we want to support custom datasets that have the `tranforms` attribute but not be properly inheriting from `VisionDataset`, we may not want to prematurely narrow the checks here.

Reply:
Any custom dataset can inherit from `VisionDataset`. This function was designed around the structure of `VisionDataset`, so I think it should be limited to datasets of that type.

Reply:
I don't think so, because Composer's synthetic datasets are of type `torch.utils.data.Dataset`, which would make them incompatible with this. But I think it could be reasonable to ensure that the dataset is a subclass of `torch.utils.data.Dataset`, which iirc is the most generic `dataset` class.

Reply:
Correct, our synthetic datasets do not support torchvision transforms, since the yield tensors directly. See https://github.com/mosaicml/composer/issues/346